### PR TITLE
feat(template): add `lookupArray` and `distinct` helpers

### PR DESCRIPTION
**Pull request by amezin**
## Changes

Add `lookupArray` and `distinct` template helpers.

## Context

I'm using `prBodyDefinitions` to link Renovate pull requests to Jira issues like this:

```js
{
  packageRules: [
    {
      matchPackageNames: ['...'],
      matchNewValue: '...',
      prBodyDefinitions: {
        Issue: 'ABC-123'
      },
      commitMessagePrefix: '{{#each upgrades}}{{{prBodyDefinitions.Issue}}}{{/each}}'
    }
  ]
}
```

But sometimes I get duplicate issue references (for example, when updating the version in multiple locations).

I thought about adding JSONata helper, but JSONata evaluation is async, and handlebars helpers must be sync. So the only option is to add multiple new helpers.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Should both of these have defensive code added in case a non-array is passed?

File: docs/usage/templates.md [hunk 1]
```diff
@@ -116,6 +116,28 @@ Returns `true` if the value is included on the list given.
 
 `{{#if (includes labels 'dependencies')}}Production Dependencies{{else}}Not Production Dependencies{{/if}}`
 
+### lookupArray
```
Comment (viceice) [inline docs/usage/templates.md:119]: we should sort them alphabetically

**Comment by amezin** [comment]
> Should both of these have defensive code added in case a non-array is passed?

Is `toArray()` not enough?

**Comment by rarkins** [comment]
What's the result if a non array is passed to it?

**Comment by amezin** [comment]
https://github.com/renovatebot/renovate/blob/main/lib/util/array.ts#L25

File: lib/util/template/index.ts [hunk 2]
```diff
@@ -87,6 +88,31 @@ handlebars.registerHelper({
...
+    toArray(obj)
+      // skip elements like #with does
+      .filter((element) => !handlebars.Utils.isEmpty(element))
+      .map((element) => options.lookupProperty(element, key))
+      .filter((value) => value !== undefined)
+  );
+});
+
+handlebars.registerHelper('distinct', (obj) => {
+  const seen = new Set();
```
Comment (viceice) [inline lib/util/template/index.ts:102]:
```suggestion
  const seen = new Set<string>();
```

File: lib/util/template/index.ts [hunk 3]
```diff
@@ -87,6 +88,31 @@ handlebars.registerHelper({
...
+  return (
+    toArray(obj)
+      // skip elements like #with does
+      .filter((element) => !handlebars.Utils.isEmpty(element))
+      .map((element) => options.lookupProperty(element, key))
+      .filter((value) => value !== undefined)
+  );
+});
+
+handlebars.registerHelper('distinct', (obj) => {
```
Comment (viceice) [inline lib/util/template/index.ts:101]:
```suggestion
handlebars.registerHelper('distinct', (obj: unknown) => {
```

File: lib/util/template/index.ts [hunk 4]
```diff
@@ -87,6 +88,31 @@ handlebars.registerHelper({
   },
 });
 
+handlebars.registerHelper('lookupArray', (obj, key, options) => {
+  return (
```
Comment (viceice) [inline lib/util/template/index.ts:92]:
```suggestion
    if (
      !obj ||
      !is.nonEmptyStringAndNotWhitespace(key) ||
      !options ||
      !('lookupProperty' in options)
    ) {
      return [];
    }

  return (
```

validation

File: lib/util/template/index.ts [hunk 5]
```diff
@@ -87,6 +88,31 @@ handlebars.registerHelper({
   },
 });
 
+handlebars.registerHelper('lookupArray', (obj, key, options) => {
```
Comment (viceice) [inline lib/util/template/index.ts:91]:
```suggestion
handlebars.registerHelper(
  'lookupArray',
  (obj: unknown, key: unknown, options: Options) => {
```

File: lib/util/template/index.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ import is from '@sindresorhus/is';
 import handlebars from 'handlebars';
```
Comment (viceice) [inline lib/util/template/index.ts:2]:
```suggestion
import handlebars, { type HelperOptions } from 'handlebars';
```

File: lib/util/template/index.ts [hunk 7]
```diff
@@ -2,6 +2,7 @@ import is from '@sindresorhus/is';
...
 import { GlobalConfig } from '../../config/global';
 import { logger } from '../../logger';
+import { toArray } from '../array';
 import { getChildEnv } from '../exec/utils';
 import { regEx } from '../regex';
```
Comment (viceice) [inline lib/util/template/index.ts:7]:
```suggestion
import { regEx } from '../regex';

// Missing in handlebars
type Options = HelperOptions & {
  lookupProperty: (element: unknown, key: unknown) => unknown;
};
```

File: lib/util/template/index.ts (hunk 4 above)
Comment (amezin) [reply lib/util/template/index.ts]: Handlebars' own `lookup` does none of this: https://github.com/handlebars-lang/handlebars.js/blob/master/lib/handlebars/helpers/lookup.js

https://handlebarsjs.com/api-reference/helpers.html `lookupProperty` isn't described as "optional" - as far as I understand, it's always passed and is part of the API.

Null key doesn't cause any error. Handlebars' `lookup` helper simply passes key to `lookupProperty` as is.

If obj is null, it'll be converted to `[null]` and then filtered out by `!handlebars.Utils.isEmpty()`.

And if I'm passing wrong number of arguments, I'd prefer an error, not just silent empty result.

Comment (viceice) [reply lib/util/template/index.ts]: add tests

Comment (amezin) [reply lib/util/template/index.ts]: I already did it: https://github.com/renovatebot/renovate/pull/30618/commits/5dba00c1629abcb546b2ef9079e5d6412bdf3a87

is anything missing?

File: lib/util/template/index.spec.ts [hunk 8]
```diff
@@ -423,4 +423,104 @@ describe('util/template/index', () => {
...
+            },
+          ],
+        },
+        false,
+      );
+
+      expect(output).toBe(' ABC-123');
+    });
+
+    it('handles null key', () => {
```
Comment (viceice) [inline lib/util/template/index.spec.ts:481]: test null object

File: lib/util/template/index.spec.ts [hunk 9]
```diff
@@ -423,4 +423,104 @@ describe('util/template/index', () => {
...
+        },
+        false,
+      );
+
+      expect(output).toBe(' ABC-123');
+    });
+  });
+
+  describe('distinct', () => {
+    it('skips duplicate values', () => {
```
Comment (viceice) [inline lib/util/template/index.spec.ts:499]: test null object

File: lib/util/template/index.spec.ts (hunk 8 above)
Comment (amezin) [reply lib/util/template/index.spec.ts:481]: There is already "handles null input array" test, isn't it doing what you're asking?

And "performs lookup for every array element" test already tests null elements.

Comment (viceice) [reply lib/util/template/index.spec.ts:481]: ok

**Review by viceice — APPROVED** [review]

**Pull request merged**